### ISSUE TITLE
Simple Qt program not working

### ISSUE THREAD

**Issue by wolfseifert** [bytedeco/javacpp-presets#1062]
The following simple qt program `Buttons.java`:
```
import org.bytedeco.javacpp.IntPointer;
import org.bytedeco.javacpp.Loader;
import org.bytedeco.javacpp.Pointer;
import org.bytedeco.javacpp.PointerPointer;
import org.bytedeco.qt.Qt5Core.QString;
import org.bytedeco.qt.Qt5Widgets.QApplication;
import org.bytedeco.qt.Qt5Widgets.QGridLayout;
import org.bytedeco.qt.Qt5Widgets.QPushButton;
import org.bytedeco.qt.Qt5Widgets.QWidget;
import org.bytedeco.qt.global.Qt5Core;
import org.bytedeco.qt.helper.Qt5Widgets.ClickedCallback;

import java.io.File;

import static org.bytedeco.qt.helper.Qt5Widgets.QAbstractButton_clicked;

public class Buttons {
  public static void main(String... args) {
    String path = Loader.load(Qt5Core.class);
    IntPointer argc = new IntPointer(new int[]{3});
    PointerPointer<Pointer> argv = new PointerPointer<>("Buttons", "-platformpluginpath", new File(path).getParent(), null);
    QApplication app = new QApplication(argc, argv);
    //new QApplication(argc, argv);
    QWidget widget = new QWidget();
    QGridLayout layout = new QGridLayout();
    int num = 0;
    for (int row = 0; row < 10; ++row)
      for (int col = 0; col < 5; ++col) {
        QPushButton button = new QPushButton();
        ++num;
        String text = "Button#" + num;
        button.setText(QString.fromStdString(text));
        ClickedCallback callback = new ClickedCallback() {
          @Override
          public void clicked(boolean checked) {
            System.out.println(text + " clicked");
          }
        };
        QAbstractButton_clicked(button, button, callback, callback.id);
        layout.addWidget(button, row, col);
      }
    widget.setLayout(layout);
    widget.show();
    QApplication.exec();
  }
}
```
and its `build.gradle.kts`:
```
plugins {
  application
}

repositories {
  mavenCentral()
}

dependencies {
  implementation("org.bytedeco:qt-platform:5.15.2-1.5.5")
}

application {
  mainClass.set("Buttons")
}
```
core-dump at runtime when pressing a button with a number higher than 10. This seems to be related to an array of size 10 of function pointers in the generated C++ code. If the button number is higher than 10 the functor is NULL which leads to a core dump.

This happens on both linux and windows and jdk 8, 11 or 16.

When removing the (unused) `QApplication app` reference variable the program exits with `QWidget: Must construct a QApplication before a QWidget` on windows. On the JVM an unused variable may be removed (IDEA suggests it), but using C++ it makes a difference (immediate destruction vs. destruction at the end of the block). In fact the corresponding C++ code needs a named variable like `auto app = QApplication{argc, argv}`. But on the JVM this is surprising.

Even more core dumps happen using jdk 8 by quickly clicking the buttons...

**Comment by saudet** [comment]
> core-dump at runtime when pressing a button with a number higher than 10. This seems to be related to an array of size 10 of function pointers in the generated C++ code. If the button number is higher than 10 the functor is NULL which leads to a core dump.
> 
> This happens on both linux and windows and jdk 8, 11 or 16.

Right, it looks like that code is using function pointers, which isn't a good idea, it should be using function objects or virtual functions, but anyway, if that's the only issue you're having, we can easily increase that to 100 or more.

/cc @freastro

> When removing the (unused) `QApplication app` reference variable the program exits with `QWidget: Must construct a QApplication before a QWidget` on windows. On the JVM an unused variable may be removed (IDEA suggests it), but using C++ it makes a difference (immediate destruction vs. destruction at the end of the block). In fact the corresponding C++ code needs a named variable like `auto app = QApplication{argc, argv}`. But on the JVM this is surprising.

Why is this surprising? Like you say, that's just how it works. To have it behave more like C++ code, we can use PointerScope: http://bytedeco.org/news/2018/07/17/bytedeco-as-distribution/
 
> Even more core dumps happen using jdk 8 by quickly clicking the buttons...

The application object might just be getting deallocated sooner. Other than PointerScope, we can make sure that doesn't happen by calling close() or something on it.

**Comment by wolfseifert** [comment]
> we can easily increase that to 100 or more

This will not solve the problem if someone else needs 101 buttons tomorrow.

> When removing the (unused) QApplication app reference variable the program exits with QWidget: Must construct a QApplication before a QWidget on windows. On the JVM an unused variable may be removed (IDEA suggests it)

In fact the [Java-Specification](https://docs.oracle.com/javase/specs/jls/se8/html/jls-12.html#jls-12.6.1) states that

> Optimizing transformations of a program can be designed that reduce the number of objects that are reachable to be less than those which would naively be considered reachable. For example, a Java compiler or code generator may choose to set a variable or parameter that will no longer be used to null to cause the storage for such an object to be potentially reclaimable sooner.

In the case above this means

```
QApplication app = new QApplication(argc, argv);
app = null;
```
which is the same as

`new QApplication(argc, argv);`

To resolve the difference in semantics between C++ and Java one could write

`app.exec();`

in the last line, but this gives a warning because `exec()` is a static method.

**Comment by saudet** [comment]
> > we can easily increase that to 100 or more
> 
> This will not solve the problem if someone else needs 101 buttons tomorrow.

Ok, so we're going to need to do something about that and get rid of the function pointers here, which shouldn't have been used in the first place...

> To resolve the difference in semantics between C++ and Java one could write
> 
> `app.exec();`
> 
> in the last line, but this gives a warning because `exec()` is a static method.

If it's a static method, it's not going to do anything to prevent the GC from reclaiming that object.

**Comment by newhinton** [comment]
@saudet How would i recompile the qt package to get more than 10 Buttons/QActions? I cant find any reference to a magic number 10.

> Right, it looks like that code is using function pointers, which isn't a good idea, it should be using function objects or virtual functions, but anyway, if that's the only issue you're having, we can easily increase that to 100 or more.

How do i do that?

**Comment by saudet** [comment]
If you're interested, for example, by `ClickedCallback`, we can annotate it with `@Allocator(max = ...)` like this:
https://github.com/bytedeco/javacpp/blob/master/src/test/java/org/bytedeco/javacpp/PointerTest.java#L54

**Comment by newhinton** [comment]
Yes, thank you! 

However, i use maven, and when i put the generated jarfiles into a local repository it doesn't work. While mavencentral provides a library that i can run, the local compiled version can be compiled but not executed. 

`Could not find jniQt5Core in class, module, and library paths.`

Investigating the jarfile, there is a libjiniQt5Core.so file, so it should be there. Do you know why there are differences?

**Comment by saudet** [comment]
Yes, that works fine. That error message can only happen when it's not found in any paths, so something is missing somewhere from your class path. If it doesn't work with Maven for some reason, add the JAR files manually to your class path.
